source/MRViewer/MRObjectHighlighter.cpp

```diff
@@ -0,0 +1,172 @@
+                featureTarget.setVisualizePropertyMask( enumValue, dimensions[i] );
+        }
+
+        featureTarget.setDecorationsColorForAllViewports( decoColorUnsel, false );
+        featureTarget.setDecorationsColorForAllViewports( decoColorSel, true );
+        featureTarget.setVisualizePropertyMask( FeatureVisualizePropertyType::Subfeatures, subfeatureMask );
+        featureTarget.setPointSize( pointSize );
+        featureTarget.setLineWidth( lineWidth );
+        featureTarget.setSubfeaturePointSize( subPointSize );
+        featureTarget.setSubfeatureLineWidth( subLineWidth );
+        featureTarget.setMainFeatureAlpha( alpha );
+        featureTarget.setSubfeatureAlphaPoints( subAlphaPoints );
+        featureTarget.setSubfeatureAlphaLines( subAlphaLines );
+        featureTarget.setSubfeatureAlphaMesh( subAlphaMesh );
+    };
+}
+
+ObjectHighlighter::ObjectHighlighter( std::vector<RuleFunc> rules )
+    : rules_( std::move( rules ) )
+{}
+
+void ObjectHighlighter::highlight( std::function<void( const std::shared_ptr<const Object>& object, ModifyFunc modify )> highlightObject )
+{
+    std::unordered_set<std::shared_ptr<Object>> visitedObjects;
+
+    auto lambda = [&]( auto& lambda, const std::shared_ptr<Object>& cur ) -> void
+    {
+        bool once = true;
+
+        highlightObject( cur, [&]( std::string name ) -> Object&
+        {
+            if ( once )
+            {
+                once = false;
+
+                visitedObjects.insert( cur );
+
+                auto [iter, isNew] = rollbackEntries_.try_emplace( cur );
+
+                if ( !isNew && iter->second.name != name )
+                {
+                    iter->second.func( *iter->first );
+                    iter->second = {};
+                    isNew = true;
+                }
+
+                if ( isNew )
+                {
+                    std::vector<RollbackFunc> objRollbackFuncs;
+                    for ( const auto& rule : rules_ )
+                    {
+                        if ( auto func = rule( *cur ) )
+                            objRollbackFuncs.push_back( std::move( func ) );
+                    }
+                    assert( !objRollbackFuncs.empty() && "ObjectHighlighter: The object passed the filter, but I don't know which information about it to save for rollback." );
+
+                    iter->second.func = [funcs = std::move(objRollbackFuncs)]( Object& target )
+                    {
+                        for ( auto& func : funcs )
+                            func( target );
+                    };
+                    iter->second.name = std::move( name );
+                }
+            }
+
+            return *cur;
+        } );
+
+        for ( const auto& child : cur->children() )
+            lambda( lambda, child );
+    };
+    lambda( lambda, SceneRoot::getSharedPtr() );
+
+    for ( auto it = rollbackEntries_.begin(); it != rollbackEntries_.end(); )
```

Comment:
Iteration order will depend on the key (address of some object), which will vary on each program run.

Reply:
Yeah, but it shouldn't matter, since the objects are independent.

Reply:
Yeah, but the objects are independent, so the order shouldn't matter.